pkg/scheduler/plugins/deviceshare/device_cache.go

```diff
@@ -0,0 +1,321 @@
+			n.deviceTotal[deviceType][minor] = make(corev1.ResourceList)
+		}
+		n.deviceFree[deviceType][minor] = quotav1.SubtractWithNonNegativeResult(
+			n.deviceTotal[deviceType][minor],
+			usedResource)
+	}
+}
+
+func (n *nodeDevice) updateDeviceUsed(deviceType schedulingv1alpha1.DeviceType, allocation *apiext.DeviceAllocation, add bool) {
+	if n.deviceUsed[deviceType] == nil {
+		n.deviceUsed[deviceType] = make(deviceResources)
+	}
+	if n.deviceUsed[deviceType][int(allocation.Minor)] == nil {
+		n.deviceUsed[deviceType][int(allocation.Minor)] = make(corev1.ResourceList)
+	}
+	if add {
+		n.deviceUsed[deviceType][int(allocation.Minor)] = quotav1.Add(
+			n.deviceUsed[deviceType][int(allocation.Minor)],
+			allocation.Resources)
+	} else {
+		n.deviceUsed[deviceType][int(allocation.Minor)] = quotav1.SubtractWithNonNegativeResult(
+			n.deviceUsed[deviceType][int(allocation.Minor)],
+			allocation.Resources)
+	}
+}
+
+func (n *nodeDevice) updateAllocateSet(allocations apiext.DeviceAllocations, pod *corev1.Pod, add bool) {
+	for deviceType, v := range allocations {
+		if v == nil || len(v) == 0 {
+			break
+		}
+		if n.allocateSet[deviceType] == nil {
+			n.allocateSet[deviceType] = make(map[types.UID]*corev1.Pod)
+		}
+		if add {
+			n.allocateSet[deviceType][pod.UID] = pod
+		} else {
+			delete(n.allocateSet[deviceType], pod.UID)
+		}
+	}
+}
+
+func (n *nodeDevice) tryAllocateDevice(podRequest corev1.ResourceList) (apiext.DeviceAllocations, []error) {
+	var allocateResult apiext.DeviceAllocations
+	var errors []error
+
+	for deviceType := range deviceResourceNames {
+		switch deviceType {
+		case schedulingv1alpha1.RDMA:
+			if !hasDeviceResource(podRequest, deviceType) {
+				break
+			}
+			if err := n.tryAllocateCommonDevice(podRequest, deviceType, allocateResult); err != nil {
+				errors = append(errors, err)
+			}
+		case schedulingv1alpha1.FPGA:
+			if !hasDeviceResource(podRequest, deviceType) {
+				break
+			}
+			if err := n.tryAllocateCommonDevice(podRequest, deviceType, allocateResult); err != nil {
+				errors = append(errors, err)
+			}
+		case schedulingv1alpha1.GPU:
+			if !hasDeviceResource(podRequest, deviceType) {
+				break
+			}
+			if err := n.tryAllocateGPU(podRequest, allocateResult); err != nil {
+				errors = append(errors, err)
+			}
+		default:
+			klog.Warningf("device type %v is not supported yet", deviceType)
+		}
+	}
+
+	return allocateResult, errors
+}
+
+func (n *nodeDevice) tryAllocateCommonDevice(podRequest corev1.ResourceList, deviceType schedulingv1alpha1.DeviceType, allocateResult apiext.DeviceAllocations) error {
+	nodeDeviceTotal := n.deviceTotal[deviceType]
+	if len(nodeDeviceTotal) <= 0 {
+		return fmt.Errorf("node does not have enough %v", deviceType)
+	}
+
+	var deviceAllocations []*apiext.DeviceAllocation
+
+	if isMultipleCommonDevicePod(podRequest, deviceType) {
+		var commonDeviceWanted int64
+		var podRequestPerCard corev1.ResourceList
+		switch deviceType {
+		case schedulingv1alpha1.RDMA:
```

Comment:
the "switch deviceType" can be saved by add a new variable of  deviceResourceType to replace apiext.KoordRDMA and  apiext.KoordFPGA